### ISSUE TITLE
[CSV] Enrolment sample file and data upload

### ISSUE THREAD

**Issue by mahalakshme** [avniproject/avni-server#838]
### For enrolment sample file:
- Id from previous system column:
        - Instead of 'Id column' add 'Id from previous system' as column with mentioned descriptor in the sample file
        - Show in the descriptor row 'Can be used to later identify the entry. Optional.'
- Also add column 'Subject Id from previous system' with mentioned descriptor row. Needed to refer the created subjects. Users will easily have this more handy when compared to having the need to export and import
- Unless the form is associated with multiple programs, do not display 'Program' column in the sample file.
- Mention as 'Optional' for the fields that are not marked mandatory in the form.
- For coded concepts, mention 'Allowed values: {}'
- For coded concepts single select, mention 'Only single value allowed.'
- For coded concepts multiple select, mention 'Format: Separate multiple values by a comma.'
- For date concepts(including Enrolment and Exit date) specify the format as 'Format: DD-MM-YYYY'
- For numeric concepts, mention 'Max value allowed:' or 'Min value allowed' if high or low absolute defined. Else mention 'Allowed values: Any number'
- For text and notes concepts, mention 'Any Text'
- For read only form elements, display "The value can be auto-calculated if not entered." Make this change for generating registration sample file as well.
- Order of descriptions:
        1. Optional
        2. Allowed values/Max value, Min value/Format
        3. Only single value allowed/Separate multiple values by a comma/Default value:
        4. The value can be auto-calculated if not entered.
- Multiple descriptions in descriptor row need to be separated by '. '(full stop followed by space) and not by bar '|'.

[Enrolment sample file Link
](https://docs.google.com/spreadsheets/d/1KNok6LAn1890_PKrWQjphgbqVdFrCD9RsS3tVeG92Pk/edit?gid=601923590#gid=601923590) - Have not entered descriptor row for all fields. 

### For enrolment data upload:
- Should process the upload even if uploaded with/without descriptor row.
- Try to download the sample file generated above, and we should be able to upload data without mentioning any additional columns. 
- Case insensitive comparison for the headers to be made. If in the headers or the form element there are leading or trailing spaces it needs to be trimmed before comparison. We have got tickets for this.
- if an entry with 'Id from previous system' already exists do not create or edit and show error 'Entry with id from previous system, 5 already present in Avni'. We currently allow to edit and has led to previous values getting erased.
- Values for 'Id from previous system' should be optional. Users should be able to upload data without mentioning this. But currently this is not possible.
- Currently the date format accepted is YYYY-MM-DD. Across all date concepts and columns like Enrolment Date and Exit date, the date format, 'DD-MM-YYYY' needs to be supported. This is inline with what is used in Avni and this is the predominant Indian date format.
- Values entered for coded concept answers need to be accepted by making case in-sensitive comparison.
- For enrolment date, show the error message: 'Enrolment date cannot be in future' when enrolment date is in the future. Same for exit date as well.

### Out of scope:
(Edit of an entry will not anymore be possible via CSV. We will support only create-not scope of this card to restrict):
- Upload of data
- The idea is to make the process easy for the most commonly used data types and iteratively improve for others as well.
- Edit of forms.
- Should show error thrown by form element rule when it is violated. 

### Inputs:
- Make it a round trip - download and upload matching 
- funding approval issues

**Comment by petmongrels** [comment]
The comments in Subject story applies to this as well. Hence move to code review comments

**Comment by mahalakshme** [comment]
@ombhardwajj Moved this AC to Out of scope: `Should show error thrown by form element rule when it is violated. ` From the code comments looked like it is not handled. Let me know anyways.

Also removed this in the descriptor row: `Should show error thrown by form element rule when it is violated. ` after today's discussion.

Feel free to let me know if you prefer updating the above as part of separate card.

**Comment by ombhardwajj** [comment]
Oh ! I have not handled anything related to enrolment data upload. 
I had a wrong notion that everything related to data upload would be handled in #833 .
I will add the necessary code once #833 is complete to avoid conflicting code.
For the same reason #839 is on hold aswell.

**Comment by mahalakshme** [comment]
@ombhardwajj then will move this card also to Hold

**Comment by petmongrels** [comment]
### Updated ACs
- program has to be provided always, like done in Subject type for Subject upload

**Comment by AchalaBelokar** [comment]
- [ ] download the enrolment sample file  and see the downloaded sample file there are mandatory flied which is not showing optional and mandatory 
Eg. I Login with dinesh@adjnpct and download the csv

[sample-ProgramEnrolment---Adolescent---Individual.csv](https://github.com/user-attachments/files/20075804/sample-ProgramEnrolment---Adolescent---Individual.csv)

- [ ] also there is child program I am able to enrol the 25 year person also there is no validation error showing.

**Comment by petmongrels** [comment]
1. fixed as part of another card
2. we don't have rule based validations so it is allowed for now

**Comment by AchalaBelokar** [comment]
@petmongrels Mandatory field will be show as mandatory

**Comment by AchalaBelokar** [comment]
- [ ] Exit enrolment is also cover in this card

**Comment by AchalaBelokar** [comment]
- 

- [ ] Login with dinesh@sncu_uat on prerelease and then I download the sample file of enrolment but I find some bug there in the enrolment form of child there is Birth Weight* (kg) (>=2.5), and Weight at the time of Discharge* (kg) (>=2.5) we are showing the specific value on data entry app but in the csv it is showing the Birth Weight allows any value and for weight also It is showing the any value 

<img width="739" alt="Image" src="https://github.com/user-attachments/assets/0dbf8ba4-b2cf-410c-b8a3-f078fcd40401" />
Pls check the header of the csv 

[sample-ProgramEnrolment---Child---Individual.csv](https://github.com/user-attachments/files/20080561/sample-ProgramEnrolment---Child---Individual.csv)

- [ ] coded concept for single select 
- [ ] For coded concepts multiple select, mention 'Format: Separate multiple values by a comma.

**Comment by AchalaBelokar** [comment]
- [ ] I login with dinesh@sncu_uat and enrolment csv uploaded with basic details then I uploaded csv with all details but individual get enrolled 2 times so there should be clear message show that already enrol individual..

**Comment by petmongrels** [comment]
these are advanced error handling. we can handle these in subsequent releases when we see the type of usage.

**Comment by mahalakshme** [comment]
@petmongrels I think above can happen in usual scenario. Since in the first upload, not all rows will pass(but some will), so they will then just re-upload the file which will result in above issue. For registration we have handled that case, - by stating something 'Individual already registered.' I missed to add as AC. But I think it is basic case need to be handled.

**Comment by dinesh2096** [comment]
If we edit the enrolment we are getting the oops error after removing the descriptor row (this is org specific happens for sncu_uat only i tried in mlduat it is not happening)

Execution id : 30292

https://github.com/user-attachments/assets/a9ce6993-4737-4284-b1b5-9a113b572fe9

**Comment by petmongrels** [comment]
this is linked to the other date of birth not getting saved issue
